tutorons/css/explain.py

```diff
@@ -36,6 +45,35 @@ def extract(self, node):
         return valid_regions
 
 
+class StylesheetSelectorExtractor(object):
+
+    def extract(self, node):
+        textfield = ''.join(map(filter_non_ascii, node.text))
+        textfield_as_list = textfield.split('\n')
+        ss_offset = 0
+        if BeautifulSoup(textfield).find_all(["style"]):
+            ss_offset = textfield.find("<style>") + len("<style>")
+            textfield = ''.join(map(filter_non_ascii, BeautifulSoup(textfield).text)).strip()
+            textfield_as_list = textfield.split('\n')
+        parser = tinycss.make_parser()
+        stylesheet = parser.parse_stylesheet(textfield)
+        valid_regions = []
+        for rule in stylesheet.rules:
+            sel = rule.selector
+            ss_start = ss_offset
+            if sel.line > 1:
+                for l in xrange(sel.line - 1):
+                    ss_start += len(textfield_as_list[l]) + 1
+            ss_start += sel.column - 1
+            sel = rule.selector.as_css()
+            if rule.declarations:
+                valid_regions.append(Region(node, ss_start, ss_start+len(sel) - 1, sel))
+        for region in valid_regions:
+            if not is_selector(region.string):
+                valid_regions.remove(region)
+        return valid_regions
+
+
```

Comment:
Would be nice to split this up into conceptual "chunks" and add brief notes about what each chunk is doing.